// hefang-js/src/datetime.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var string_1 = require("./string");
var smallMonth = [4, 6, 9, 11];
function now() {
    return new Date;
}
exports.now = now;
function firstWeek(month, year) {
    var date = new Date(year, month - 1, 1);
    return date.getDay();
}
exports.firstWeek = firstWeek;
function formatDate(format, date) {
    if (format instanceof Date) {
        date = format;
        format = null;
    }
    format = format || "yyyy-MM-dd HH:mm:ss";
    date = date || now();
    var cfg = {
        yyyy: date.getFullYear(),
        yy: (date.getFullYear() + '').substring(2),
        M: date.getMonth() + 1,
        MM: string_1.padStart(date.getMonth() + 1, 2, 0),
        d: date.getDate(),
        dd: string_1.padStart(date.getDate(), 2, 0),
        h: date.getHours(),
        HH: string_1.padStart(date.getHours(), 2, 0),
        m: date.getMinutes(),
        mm: string_1.padStart(date.getMinutes(), 2, 0),
        s: date.getSeconds(),
        ss: string_1.padStart(date.getSeconds(), 2, 0)
    };
    return format.replace(/([a-z])(\1)*/ig, function (m) {
        return cfg[m] || m;
    });
}
exports.formatDate = formatDate;
function daysOfMonth(month, year) {
    return month === 2 ? (isLeapYear(year || now().getFullYear()) ? 29 : 28) :
        (smallMonth.indexOf(month) > -1 ? 30 : 31);
}
exports.daysOfMonth = daysOfMonth;
function isLeapYear(year) {
    return year % 4 === 0 && year % 100 !== 0 || year % 400 === 0;
}
exports.isLeapYear = isLeapYear;
function currentMonth() {
    var n = now();
    return {
        start: new Date(n.getFullYear(), n.getMonth(), 1, 0, 0, 0, 0),
        end: new Date(n.getFullYear(), n.getMonth(), daysOfMonth(n.getMonth() + 1), 23, 59, 59, 999)
    };
}
exports.currentMonth = currentMonth;
function lastMonth() {
    var n = now();
    return {
        start: new Date(n.getFullYear(), n.getMonth() - 1, 1, 0, 0, 0, 0),
        end: new Date(n.getFullYear(), n.getMonth() - 1, daysOfMonth(n.getMonth()), 23, 59, 59, 999)
    };
}
exports.lastMonth = lastMonth;
function today() {
    var n = now();
    return {
        start: new Date(n.getFullYear(), n.getMonth(), n.getDate(), 0, 0, 0, 0),
        end: new Date(n.getFullYear(), n.getMonth(), n.getDate(), 23, 59, 59, 999)
    };
}
exports.today = today;
function yesterday() {
    var n = now();
    return {
        start: new Date(n.getFullYear(), n.getMonth(), n.getDate() - 1, 0, 0, 0, 0),
        end: new Date(n.getFullYear(), n.getMonth(), n.getDate() - 1, 23, 59, 59, 999)
    };
}
exports.yesterday = yesterday;
